documentation/src/docs/recovery-procedures/table-segment-recovery.md

```diff
@@ -0,0 +1,185 @@
+
+# Issue
+One of the issues recently seen in Pravega was some data corruption happening on Attribute index segment. The issue manifested with the below error message popping up in logs:-
+```
+ERROR i.p.s.s.h.h.PravegaRequestProcessor - [requestId=2205414166888448] Error (Segment = '_system/_tables/completedTransactionsBatch-0', Operation = 'updateTableEntries')
+io.pravega.segmentstore.server.DataCorruptionException: BTreeIndex operation failed. Index corrupted.
+...
+Caused by: io.pravega.common.util.IllegalDataFormatException: [AttributeIndex[3-423]] Wrong footer information. RootPage Offset (3842188288) + Length (142858) exceeds Footer Offset (15007837).
+```
+
+The attribute index segment is basically a BTree Index of the data(key/value pair) that goes in its associated main Table Segment. It is internally a B+Tree (existing on Pravega segment), and is organized as as set of pages written to storage. Each page can be an index page or leaf page (holding data entries) having tree nodes. After writing a collection of pages we write something called as a footer that basically points to latest location of the root. The error above basically indicates the fact that  since footer exists at the highest offsets of every write, the root page's length from its offset cannot exceed that of footer's. The issue having happened so, indicates some kind of corruption in the underlying attribute index segment.
+
+
+# Root Cause
+
+The above issue has been seen in conjuntion with a CacheFullException. That is we see a cacheFullException first and couple of minutes later is when we see the DataCorruptionException described above. Although we have not been able to do a exact root cause analysis on why such a corruption occured we suspect it could be linked to us getting
+a cacheFullException. That is the BTree index pages accessed are also cached in Pravega's internal cache. And trying to access these pages when the cache is full could have led to some wrong data being read, and eventually wrong data being written to the underlying Pravega segments. 
+It could also be an issue pertaining to the underlying tier-2 storage that Pravega uses. That is some kind of corruption having occured in the storage system that Pravega uses that could have caused such an exception.
+However we do not know the exact root cause of what might have led to this and that is the reason this Recovery procudure exists.
+
+
+
+# Repair Procedure
+
+The repair procedure aims at recovering Pravega from such a situation. Here is what the recovery procedure aims to do. Recall that the corruption that we see occurs in the attribute index segment which is just an index of the data that exists in the associated primary table segment( holding key-value pairs ) and that this primary table segment is perfectly fine. So what if we re-read this main table segment and generate an index again which is clean. Thats what the repair procedure does. It revolves around re-reading 
```

Comment:
After the previous paragraph, we could create a new paragraph summarizing the approach. I have done a slight change to the text:
`So what if we re-read this main table segment and generate an index again which is clean. Thats what the repair procedure does. It revolves around re-reading the key values pairs in the main table segment, feeding them to a in-process Pravega. Doing so would result in an attribute index segment generated into the storage directory that this in-process Pravega points to. We replace this generated clean attribute index segment back to the Pravega cluster which had the datacoruption, therby freeing it from the issue.`
->
`In a nutshell, the recovery procedure of a Table Segment Attribute Index involves re-reading the key/value pairs in the primary Table Segment and feeding them to a local in-process Pravega cluster. Doing so would result in an Table Segment Attribute Index generated in the storage directory of that local in-process Pravega cluster. To do so, we need to collect for the corrupted Table Segment Attribute Index all the primary chunks from the Pravega cluster to repair. Then, we use the Admin CLI "dataRecovery tableSegment-recovery" command (see #6753) to re-create locally the new Table Segment Attribute Index. Finally, we need to replace the newly generated Table Segment Attribute Index by the corrupted one in the Pravega cluster to repair.`

And then, provide the detailed explanation step by step.

Reply:
done.